sentry-android-replay/src/main/java/io/sentry/android/replay/ReplayIntegration.kt

```diff
@@ -0,0 +1,317 @@
+import io.sentry.transport.ICurrentDateProvider
+import io.sentry.util.FileUtils
+import java.io.Closeable
+import java.io.File
+import java.util.Date
+import java.util.concurrent.ExecutorService
+import java.util.concurrent.Executors
+import java.util.concurrent.ThreadFactory
+import java.util.concurrent.TimeUnit.MILLISECONDS
+import java.util.concurrent.atomic.AtomicBoolean
+import java.util.concurrent.atomic.AtomicInteger
+import java.util.concurrent.atomic.AtomicReference
+import kotlin.LazyThreadSafetyMode.NONE
+import kotlin.math.roundToInt
+
+class ReplayIntegration(
+    private val context: Context,
+    private val dateProvider: ICurrentDateProvider
+) : Integration, Closeable, ScreenshotRecorderCallback {
+
+    companion object {
+        const val VIDEO_SEGMENT_DURATION = 5_000L
+        const val VIDEO_BUFFER_DURATION = 30_000L
+    }
+
+    private lateinit var options: SentryOptions
+    private var hub: IHub? = null
+    private var recorder: WindowRecorder? = null
+    private var cache: ReplayCache? = null
+
+    // TODO: probably not everything has to be thread-safe here
+    private val isEnabled = AtomicBoolean(false)
+    private val isRecording = AtomicBoolean(false)
+    private val currentReplayId = AtomicReference<SentryId>()
+    private val segmentTimestamp = AtomicReference<Date>()
+    private val currentSegment = AtomicInteger(0)
+    private val saver =
+        Executors.newSingleThreadScheduledExecutor(ReplayExecutorServiceThreadFactory())
+
+    private val screenBounds by lazy(NONE) {
+        // PixelCopy takes screenshots including system bars, so we have to get the real size here
+        val wm = context.getSystemService(Context.WINDOW_SERVICE) as WindowManager
+        if (VERSION.SDK_INT >= VERSION_CODES.R) {
+            wm.currentWindowMetrics.bounds
+        } else {
+            val screenBounds = Point()
+            @Suppress("DEPRECATION")
+            wm.defaultDisplay.getRealSize(screenBounds)
+            Rect(0, 0, screenBounds.x, screenBounds.y)
+        }
+    }
+
+    private val aspectRatio by lazy(NONE) {
+        screenBounds.bottom.toFloat() / screenBounds.right.toFloat()
+    }
+
+    private val recorderConfig by lazy(NONE) {
+        ScreenshotRecorderConfig(
+            recordingWidth = (720 / aspectRatio).roundToInt(),
+            recordingHeight = 720,
+            scaleFactor = 720f / screenBounds.bottom
+        )
+    }
+
+    override fun register(hub: IHub, options: SentryOptions) {
+        this.options = options
+
+        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.O) {
+            options.logger.log(INFO, "Session replay is only supported on API 26 and above")
+            return
+        }
+
+        // TODO: check for replaysSessionSampleRate and replaysOnErrorSampleRate
+
+        this.hub = hub
+        recorder = WindowRecorder(options, recorderConfig, this)
+        isEnabled.set(true)
+    }
+
+    fun isRecording() = isRecording.get()
+
+    fun start() {
```

Comment:
We probably should have some overall lifecycle state to guard against random invocations. E.g. pause() is ignored unless the current state is `resumed`.

Reply:
great idea! For now, we won't expose any public API to the customers, but once we do, I'll make sure to add the lifecycle state and checks, will leave a TODO as well